### ISSUE TITLE
ObjectField implementation missing in DocumentSerializer

### ISSUE THREAD

**Issue by Antman261** [barseghyanartur/django-elasticsearch-dsl-drf#11]
I've had a look at DocumentSerializer and can't find any serialization for elasticsearch-dsl's [ObjectField](https://github.com/sabricot/django-elasticsearch-dsl/blob/master/django_elasticsearch_dsl/fields.py#L80) representation of [Object](https://www.elastic.co/guide/en/elasticsearch/reference/current/object.html)

I've had a look at the implementation to see if I could come up with a quick solution but I'm not readily seeing anything.

Let me know if there's an easy way to do this or if it needs some work then point me in the right direction and I'll try to submit a PR.

**Comment by barseghyanartur** [comment]
@Antman261:

Could you describe the whole use-case?

In general, in terms of DRF, ``ObjectField`` is likely to be used for representing foreign relations (nested objects). There are many ways of dealing with it. 

So, please, describe your use case as much as possible.

**Comment by Antman261** [comment]
Here's the document

```python

@INDEX.doc_type
class DishDocument(DocType, BaseDocumentMixin):
    pg_id = fields.IntegerField(attr='id')
    keywords = fields.ListField(fields.StringField())
    likes_count = fields.IntegerField()
    users = fields.ListField(fields.IntegerField())
    image_url = fields.StringField()
    restaurant = fields.ObjectField(properties={
        'pg_id': fields.StringField(attr='id'),
        'name': fields.StringField(),
        'image_url': fields.StringField(),
        'instagram_user': fields.StringField(),
        'address': fields.StringField(),
        'cuisines': fields.ListField(fields.StringField(attr='cuisines_list_elasticsearch')),
        'information': fields.TextField(),
        'phone_number': fields.StringField(),
        'time_offset_minutes': fields.IntegerField(),
        'time_offset_hours': fields.FloatField(),
        'tripadvisor_widget': fields.TextField(),
        'latitude': fields.FloatField(),
        'longitude': fields.FloatField(),
        'highlights': fields.ListField(fields.StringField(attr='highlights_list_elasticsearch')),
        'suburb': fields.StringField(),
        'location': fields.GeoPointField(attr='location_elasticsearch'),
        'app_opening_times': fields.ObjectField(),
        'blogs': fields.NestedField(properties={
            'author': fields.StringField(),
            'image_url': fields.StringField(),
            'title': fields.StringField(),
            'url': fields.StringField()
        }),
        'delivery_type': fields.StringField(attr='delivery_provider.slug'),
        'delivery_link': fields.StringField(),
        'quandoo_id': fields.IntegerField()
    })
    price = fields.IntegerField()
    description = fields.TextField()

    class Meta:
        model = Dish
        fields = ['title', 'instagram_user']
```

And the models:

```python
class Restaurant(Creatable):
    name = models.CharField(max_length=255)
    slug = models.SlugField(default='', unique=True)
    image_url = models.URLField(max_length=600)
    address = models.TextField(blank=True, default='')
    cuisines = models.ManyToManyField(Cuisine)
    information = models.TextField(blank=True, default='')
    highlights = models.ManyToManyField(Highlight)
    blogs = models.ManyToManyField(Blog, related_name='restaurant')
    phone_number = models.CharField(max_length=20, blank=True, default='')
    suburb = models.CharField(max_length=55, blank=True, default='')
    instagram_user = models.CharField(max_length=61, blank=True, default='')
    time_offset_minutes = models.IntegerField(help_text="Multiply hours by 60",
                                              default=0)
    timezone = TimeZoneField(default='', blank=True)
    tripadvisor_widget = models.TextField(blank=True, default='')
    location = gis_models.PointField(default="POINT(0.0 0.0)", blank=True, db_index=True)
    latitude = models.DecimalField(max_digits=10, decimal_places=7, default=0)
    longitude = models.DecimalField(max_digits=10, decimal_places=7, default=0)
    firebase_id = models.CharField(max_length=255, default='', blank=True, unique=True)
    quandoo_id = models.BigIntegerField(blank=True, null=True)
    delivery_provider = models.ForeignKey('DeliveryProvider', on_delete=models.SET_NULL,
                                          null=True, blank=True)
    delivery_link = models.URLField(max_length=512, blank=True, null=True)


class Dish(Creatable):
    restaurant = models.ForeignKey(Restaurant, on_delete=models.CASCADE,
                                   related_name='dishes')
    image_url = models.URLField(blank=True, default='', max_length=600)
    price = models.IntegerField(default=0)
    title = models.CharField(max_length=600)
    description = models.TextField(default='', blank=True)
    recipe = models.ForeignKey('Recipe', on_delete=models.SET_NULL, null=True,
                               related_name='dishes', blank=True)
    views_count = models.PositiveIntegerField(
        default=0,
        help_text='legacy calculated field from Firebase'
    )
    likes_count = models.PositiveIntegerField(
        default=0,
        help_text='legacy calculated field from Firebase'
    )
    instagram_user = models.CharField(max_length=61, blank=True, default='')
    keywords = models.ManyToManyField(Keyword)
    tags = models.ManyToManyField(Tag)
    firebase_id = models.CharField(max_length=255, default='', blank=True, unique=True)
    random = models.BigIntegerField(default=random_number)

```

When I query elastic search directly I get the correct data back

```json

{
  "took" : 73,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 54,
    "max_score" : 26.304188,
    "hits" : [
      {
        "_index" : "dishes",
        "_type" : "dish_document",
        "_id" : "2351",
        "_score" : 26.304188,
        "_source" : {
          "price" : 1500,
          "users" : [
            19301,
            19562,
            21291,
            20211,
            17843,
          ],
          "title" : "Vegan curry pie",
          "restaurant" : {
            "app_opening_times" : [
              { },
              { },
              { },
              { },
              { },
              { }
            ],
            "tripadvisor_widget" : "",
            "name" : "Henry Saw",
            "highlights" : [
              "Breakfast",
              "Takeaway Available",
              "Outdoor Seating"
            ],
            "information" : "Vegan & plant based friendly including wraps, pies, sweet treats, smoothies, tea and even Kombucha.",
            "blogs" : [ ],
            "delivery_type" : null,
            "suburb" : "Perth CBD",
            "pg_id" : 2878,
            "quandoo_id" : null,
            "address" : "117 Grand Lane, Perth",
            "delivery_link" : null,
            "longitude" : 115.89077,
            "time_offset_hours" : 8.0,
            "cuisines" : [
              "Cafe"
            ],
            "location" : {
              "lat" : -31.91389,
              "lon" : 115.89077
            },
            "latitude" : -31.91389,
            "instagram_user" : "",
            "time_offset_minutes" : 480,
            "image_url" : "https://cdn.feedmeeapp.com/images/1488647004249.jpg",
            "phone_number" : "0404545729"
          },
          "likes_count" : 257,
          "keywords" : [
            "lunch",
            "vegan",
            "breakfast",
            "vegetarian"
          ],
          "description" : "with salad medley",
          "pg_id" : 2351,
          "image_url" : "https://cdn.feedmeeapp.com/images/1472878675380.jpg",
          "instagram_user" : "shanmintay"
        }
      },
```

**Comment by barseghyanartur** [comment]
@Antman261:

There isn't an out-of-box solution for that (yet), except that I recommend doing something like this in your serializer:

```python

from rest_framework import serializers
from django_elasticsearch_dsl_drf.serializers import DocumentSerializer

from .documents import DishDocument

class DishDocumentSerializer(DocumentSerializer):

    restaurant = serializers.SerializerMethodField()

    class Meta(object):
        """Meta options."""

        document = DishDocument
        fields = (
            'pg_id',
            'keywords',
            'likes_count',
            'users',
            'image_url',
            'restaurant',
            'price',
            'description',
        )

    def get_restaurant(self, obj):
        if obj.restaurant:
            return dict(obj.restaurant)
        else:
            return {}

```

**Comment by barseghyanartur** [comment]
@Antman261:

Some progress has been made to address the issue. Check the current master:

- [sample model](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/examples/simple/books/models/address.py)
- [sample related model](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/examples/simple/books/models/city.py)
- [document](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/examples/simple/search_indexes/documents/address.py)
- [serializer](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/examples/simple/search_indexes/serializers/address.py)
- [view-set](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/examples/simple/search_indexes/viewsets/address.py)

Nesting one level deep seems to work fine.

I haven't tested it thoroughly yet though. Perhaps you could help.

**Comment by Antman261** [comment]
With the above code I get this error:

```
  File "/Users/anthonymanning-franklin/repos/Feedmee/fm-django/env/lib/python3.4/site-packages/django_elasticsearch_dsl_drf/serializers.py", line 267, in get_fields
    self._field_mapping[field_type.__class__](**kwargs)
KeyError: <class 'django_elasticsearch_dsl.fields.ObjectField'>
```

If I add a mapping like this

```python
class DocumentSerializer(
    six.with_metaclass(DocumentSerializerMeta, serializers.Serializer)
):
    """A dynamic DocumentSerializer class."""

    _abstract = True

    _field_mapping = {
        # models.AutoField: IntegerField,
        fields.AttachmentField: CharField,  # TODO
        fields.BooleanField: BooleanField,
        fields.ByteField: CharField,  # TODO
        # fields.CompletionField: ,  # TODO
        fields.DateField: DateField,
        fields.DoubleField: FloatField,
        fields.FloatField: FloatField,
        fields.ObjectField: SerializerMethodField,
```

I get this error

```
  File "/Users/anthonymanning-franklin/repos/Feedmee/fm-django/foodrank/serializers.py", line 38, in get_restaurant
    return dict(obj.restaurant)
ValueError: dictionary update sequence element #0 has length 19; 2 is required
```

In the end I wrote another serializer subclassing ModelSerializer and that worked,  but it was dependent on adding an element for `fields.ObjectField` in `_field_mapping`.

**Comment by Antman261** [comment]
I replaced my pypi version with the master branch and it worked perfectly, using a ModelSerializer for the ObjectField.

**Comment by barseghyanartur** [comment]
@Antman261:

Good to know. Changes will be released in version 0.4.5. Until then, keep using master.

**Comment by barseghyanartur** [comment]
This needs better testing still, since it now fails on detail views.

**Comment by barseghyanartur** [comment]
@Antman261:

It should now (master) work well out of the box with absolutely no customisations in document serializers (the dynamic ones). Could you check if it works in your case?

**Comment by barseghyanartur** [comment]
Solved in [0.5](https://pypi.python.org/pypi/django-elasticsearch-dsl-drf/0.5).
